File: packages/contracts/src/semantics.ts
import { selectedCaptureMasterProfile } from "./capture-master-profile.ts";
import type { CallDocument } from "./document-schema.ts";
import { frameCountForDuration, selectedMediaProfile, waveByteLength } from "./media-profile.ts";
export function requireValid(condition: unknown): asserts condition {
  if (!condition) {
    throw new Error("semantics");
  }
}
export function unique(values: readonly unknown[]): void {
  requireValid(new Set(values).size === values.length);
}
export function validateCall(call: CallDocument): void {
  unique(call.tracks.map((t) => t.trackId));
  unique(call.tracks.map((t) => t.role));
  const finalized = call.captureState !== "recording";
  requireValid(
    finalized
      ? call.durationMs !== null && call.endedAt !== null
      : call.durationMs === null && call.endedAt === null,
  );
  if (call.endedAt !== null) {
    requireValid(Date.parse(call.endedAt) >= Date.parse(call.startedAt));
  }
  requireValid(
    call.captureState === "interrupted"
      ? call.interruptionReason !== null
      : call.interruptionReason === null,
  );
  for (const track of call.tracks) {
    let cursor = 0;
    for (const interval of track.intervals) {
      requireValid(interval.endMs > interval.startMs && interval.startMs === cursor);
      requireValid(interval.state !== "muted" || track.role === "microphone");
      cursor = interval.endMs;
    }
    if (finalized) {
      requireValid(cursor === call.durationMs);
    }
  }
  unique(call.revisions.map((r) => r.revisionId));
  requireValid(
    call.activeRevisionId === null ||
      call.revisions.some((r) => r.revisionId === call.activeRevisionId),
  );
  const groupIds: string[] = [];
  for (const [revisionId, groups] of Object.entries(call.speakerGroups)) {
    requireValid(call.revisions.some((revision) => revision.revisionId === revisionId));
    const members: string[] = [];
    for (const group of groups) {
      groupIds.push(group.groupId);
      members.push(...group.speakerIds);
      unique(group.speakerIds);
    }
    unique(members);
  }
  unique(groupIds);
}
export function validateRevision(
  revision: import("./document-schema.ts").TranscriptRevision,
): void {
  unique(revision.speakers.map((s) => s.speakerId));
  unique(revision.turns.map((t) => t.turnId));
  let previousStart = 0;
  const previousTrackStarts = new Map<string, number>();
  for (const turn of revision.turns) {
    requireValid(turn.endMs >= turn.startMs);
    if (revision.normalizationVersion === 1) {
      requireValid(turn.startMs >= previousStart);
      previousStart = turn.startMs;
    } else if (!turn.words.some((word) => word.timingUncertain === true)) {
      requireValid(turn.startMs >= (previousTrackStarts.get(turn.trackId) ?? 0));
      previousTrackStarts.set(turn.trackId, turn.startMs);
    }
    if (turn.speakerId !== null) {
      requireValid(
        revision.speakers.some((s) => s.speakerId === turn.speakerId && s.trackId === turn.trackId),
      );
    }
    let cursor = turn.startMs;
    for (const word of turn.words) {
      requireValid(revision.normalizationVersion === 2 || word.timingUncertain === undefined);
      if (word.timingUncertain === true) {
        continue;
      }
      requireValid(
        word.startMs >= cursor && word.endMs >= word.startMs && word.endMs <= turn.endMs,
      );
      cursor = word.endMs;
    }
  }
}
export function validateAudio(audio: import("./document-schema.ts").AudioManifest): void {
  const master = audio.mediaProfileId === selectedCaptureMasterProfile.id;
  requireValid(master || audio.mediaProfileId === selectedMediaProfile.id);
  if (master) {
    requireValid(
      audio.durationMs <= selectedCaptureMasterProfile.maxCallDurationMs &&
        audio.objects.length === (audio.durationMs === 0 ? 0 : 1),
    );
  }
  unique(audio.objects.map((o) => o.objectId));
  unique(audio.objects.map((o) => o.index));
  let index = -1;
  let start = 0;
  for (const object of audio.objects) {
    requireValid(
      object.index > index &&
        object.startMs >= start &&
        object.endMs > object.startMs &&
        object.endMs <= audio.durationMs,
    );
    index = object.index;
    start = object.startMs;
    unique(object.channelMap.map((c) => c.channelIndex));
    unique(object.channelMap.map((c) => c.trackId));
    const durationMs = object.endMs - object.startMs;
    requireValid(
      (master
        ? object.contentType === selectedCaptureMasterProfile.contentType &&
          object.index === 0 &&
          object.startMs === 0 &&
          object.endMs === audio.durationMs &&
          object.byteLength === selectedCaptureMasterProfile.headerBytes + durationMs * 64
        : object.contentType === selectedMediaProfile.contentType &&
          durationMs <= selectedMediaProfile.objectDurationMs &&
          object.byteLength === waveByteLength(frameCountForDuration(durationMs)) &&
          object.byteLength <= selectedMediaProfile.maxObjectBytes &&
          object.byteLength <= selectedMediaProfile.limits.uploadRequestBytes) &&
        object.channelMap.length === selectedMediaProfile.channels.length &&
        selectedMediaProfile.channels.every((expected) =>
          object.channelMap.some((channel) => channel.channelIndex === expected.index),
        ),
    );
  }
}

File: packages/contracts/Sources/TrigoContracts/Semantics.swift
import Foundation
import JSONSchema

extension Contract {
  static func validateCall(_ call: JSONValue) throws {
    let tracks = call["tracks"].items
    try unique(tracks.map { $0["trackId"] })
    try unique(tracks.map { $0["role"] })
    let finalized = call["captureState"].text != "recording"
    try require(
      finalized
        ? !call["durationMs"].isNull && !call["endedAt"].isNull
        : call["durationMs"].isNull && call["endedAt"].isNull)
    if !call["endedAt"].isNull {
      func date(_ text: String) -> Date? {
        let f = ISO8601DateFormatter()
        if text.contains(".") { f.formatOptions.insert(.withFractionalSeconds) }
        return f.date(from: text)
      }
      if let start = date(call["startedAt"].text), let end = date(call["endedAt"].text) {
        try require(end >= start)
      } else {
        throw ContractError.semantics
      }
    }
    try require(
      call["captureState"].text == "interrupted"
        ? !call["interruptionReason"].isNull : call["interruptionReason"].isNull)
    for track in tracks {
      var cursor = 0
      for interval in track["intervals"].items {
        try require(
          interval["endMs"].integerValue > interval["startMs"].integerValue
            && interval["startMs"].integerValue == cursor)
        try require(interval["state"].text != "muted" || track["role"].text == "microphone")
        cursor = interval["endMs"].integerValue
      }
      if finalized { try require(cursor == call["durationMs"].integerValue) }
    }
    try unique(call["revisions"].items.map { $0["revisionId"] })
    try require(
      call["activeRevisionId"].isNull
        || call["revisions"].items.contains { $0["revisionId"] == call["activeRevisionId"] })
  }
  static func validateRevision(_ revision: JSONValue) throws {
    let speakers = revision["speakers"].items
    let turns = revision["turns"].items
    try unique(speakers.map { $0["speakerId"] })
    try unique(turns.map { $0["turnId"] })
    var previousStart = 0
    for turn in turns {
      let start = turn["startMs"].integerValue
      let end = turn["endMs"].integerValue
      try require(start >= previousStart && end >= start)
      previousStart = start
      if !turn["speakerId"].isNull {
        try require(
          speakers.contains {
            $0["speakerId"] == turn["speakerId"] && $0["trackId"] == turn["trackId"]
          })
      }
      var cursor = start
      for word in turn["words"].items {
        try require(
          word["startMs"].integerValue >= cursor
            && word["endMs"].integerValue >= word["startMs"].integerValue
            && word["endMs"].integerValue <= end)
        cursor = word["endMs"].integerValue
      }
    }
  }
  static func validateAudio(_ audio: JSONValue) throws {
    let profile = try MediaProfile.selected()
    try require(audio["mediaProfileId"].text == profile.id.rawValue)
    let objects = audio["objects"].items
    try unique(objects.map { $0["objectId"] })
    try unique(objects.map { $0["index"] })
    var index = -1
    var start = 0
    for object in objects {
      try require(
        object["index"].integerValue > index && object["startMs"].integerValue >= start
          && object["endMs"].integerValue > object["startMs"].integerValue
          && object["endMs"].integerValue <= audio["durationMs"].integerValue)
      index = object["index"].integerValue
      start = object["startMs"].integerValue
      try unique(object["channelMap"].items.map { $0["channelIndex"] })
      try unique(object["channelMap"].items.map { $0["trackId"] })
      let durationMs = object["endMs"].integerValue - object["startMs"].integerValue
      try require(durationMs <= profile.objectDurationMs)
      let frameCount = profile.frameCount(durationMs: durationMs)
      try require(
        object["contentType"].text == profile.contentType.rawValue
          && object["byteLength"].integerValue == profile.waveByteLength(frameCount: frameCount)
          && object["byteLength"].integerValue <= profile.maxObjectBytes
          && object["byteLength"].integerValue <= profile.limits.uploadRequestBytes
          && object["channelMap"].items.count == profile.channels.count
          && profile.channels.allSatisfy { expected in
            object["channelMap"].items.contains {
              $0["channelIndex"].integerValue == expected.index
            }
          })
    }
  }
}
